kernel: ASSERTION FAIL [z_spin_lock_valid(l)]

This is similar to #13289 but 100% reproducible with a simple `main.c`

Run the following on `qemu_x86` or any `nrf` board:
 
https://github.com/lemrey/zephyr/tree/bug-workq

```
prj.conf:
CONFIG_ASSERT=y
---
main.c:
#include <zephyr.h>
#include <misc/printk.h>

struct k_delayed_work dw;

void work_handler(struct k_work *w)
{
	k_delayed_work_submit(&dw, 10);
}

void main(void)
{
	printk("Hello World! %s\n", CONFIG_BOARD);

	k_delayed_work_init(&dw, work_handler);
	k_delayed_work_submit(&dw, K_NO_WAIT);

	while (1) {
		k_sleep(K_SECONDS(1));
	}
}
```

results in:
`ASSERTION FAIL [z_spin_lock_valid(l)] @ zephyr.git/include/spinlock.h:66`

@andyross thanks for looking into this. I tried reverting a37a981, but:

```
[8/98] Building C object zephyr/CMakeFiles/offsets.dir/arch/x86/core/offsets/offsets.c.obj
FAILED: zephyr/CMakeFiles/offsets.dir/arch/x86/core/offsets/offsets.c.obj
/home/carles/bin/zephyr-sdk/sysroots/x86_64-pokysdk-linux/usr/bin/i586-zephyr-elf/i586-zephyr-elf-gcc -DBUILD_VERc
In file included from ../../../kernel/include/kernel_structs.h:10:0,
                 from /home/carles/src/zephyr/zephyr/arch/x86/core/offsets/offsets.c:29:
../../../include/kernel.h:1735:20: error: field ‘lock’ has incomplete type
  struct k_spinlock lock;
                    ^~~~
../../../include/kernel.h:2377:20: error: field ‘lock’ has incomplete type
  struct k_spinlock lock;
                    ^~~~
../../../include/kernel.h:3154:20: error: field ‘lock’ has incomplete type
  struct k_spinlock lock;
                    ^~~~
../../../include/kernel.h:3457:20: error: field ‘lock’ has incomplete type
  struct k_spinlock lock;
                    ^~~~
../../../include/kernel.h:3637:20: error: field ‘lock’ has incomplete type
  struct k_spinlock lock;  /**< Synchronization lock */
                    ^~~~
ninja: build stopped: subcommand failed.
```

--- a/kernel/work_q.c
+++ b/kernel/work_q.c
@@ -74,7 +74,7 @@ int k_delayed_work_submit_to_queue(struct k_work_q *work_q,
 				   s32_t delay)
 {
 	k_spinlock_key_t key = k_spin_lock(&work_q->lock);
-	int err;
+	int err = 0;
 
 	/* Work cannot be active in multiple queues */
 	if (work->work_q && work->work_q != work_q) {
@@ -93,20 +93,21 @@ int k_delayed_work_submit_to_queue(struct k_work_q *work_q,
 	/* Attach workqueue so the timeout callback can submit it */
 	work->work_q = work_q;
 
+	/* Submit work directly if no delay.  Note that this is a
+	 * blocking operation, so release the lock first.
+	 */
 	if (!delay) {
-		/* Submit work if no ticks is 0 */
+		k_spin_unlock(&work_q->lock, key);
 		k_work_submit_to_queue(work_q, &work->work);
-	} else {
-		/* Add timeout */
-		_add_timeout(&work->timeout, work_timeout,
-			     _TICK_ALIGN + _ms_to_ticks(delay));
+		return 0;
 	}
 
-	err = 0;
+	/* Add timeout */
+	_add_timeout(&work->timeout, work_timeout,
+		     _TICK_ALIGN + _ms_to_ticks(delay));
 
 done:
 	k_spin_unlock(&work_q->lock, key);
-
 	return err;
 }
 

